src/GCodes/GCodes2.cpp

```diff
@@ -2381,6 +2381,98 @@ bool GCodes::HandleMcode(GCodeBuffer& gb, const StringRef& reply)
 		result = SetHeaterModel(gb, reply);
 		break;
 
+	case 308: // Set generic heater state
+	{
+		/*
+		 * M308 Generic Heater Control
+		 * 	Pnnn Heater number
+		 * 	Snnn Active Temp
+		 * 	Rnnn Standby Temp
+		 * 	Tnnn State: 0 = off, 1 = standby 2 = on.
+		 */
+		Heat& heat = reprap.GetHeat();
+		bool seen = false;
+		uint8_t currentHeater = 0;
+
+		if (gb.Seen('P')) {
+			currentHeater = gb.GetUIValue();
+			if (currentHeater < 0 || currentHeater >= NumHeaters)
+			{
+				reply.printf("Invalid heater index '%d'", currentHeater);
+				result = GCodeResult::error;
+				break;
+			}
+		} else {
+			reply.copy("P parameter must be specified");
+			result = GCodeResult::error;
+			break;
+		}
+
+		const Heat::HeaterStatus currentHeaterStatus = heat.GetStatus(currentHeater);
+		if (currentHeaterStatus > Heat::HS_active) {
+			reply.printf("Heater is currently in %s state",
+				heat.HeaterStatusToString(currentHeaterStatus));
+			result = GCodeResult::error;
+			break;
+		}
+
+		// Active temperature
+		if (gb.Seen('S'))
+		{
+			seen = true;
+			heat.SetActiveTemperature(currentHeater, gb.GetFValue());
+		}
+
+		if (gb.Seen('R'))
+		{
+			seen = true;
+			heat.SetStandbyTemperature(currentHeater, gb.GetFValue());
+		}
+
+		if (gb.Seen('T'))
+		{
+			seen = true;
+			const int8_t state = gb.GetIValue();
+			if (state < Heat::HS_off || state > Heat::HS_active) {
+				reply.copy("T parameter must be 0(off), 1(standby) or 2(active)");
+				result = GCodeResult::error;
+				break;
+			}
+
+			Tool* tool = reprap.GetCurrentTool();
+			ToolState toolState = tool->GetState();
+			if (toolState == ToolState::active && tool->IsHeaterAssignedToTool(currentHeater)) {
```

Comment:
Wouldn't it be more convenient here to check with `reprap.GetPrintMonitor().IsPrinting()` and `GCode::IsPaused()` for the printing state?
My reasoning behind this is that otherwise one cannot disable the heater with this command if only one tool is defined and that has been activated via `T0` at the end of `config.g`.

Reply:
I went back and forth on this.  If I allow the state change while not printing, then the tool can be active and think the heater is active when the heater itself isn't.   Let me see what happens to the DWC and PanelDue in this state.